redhat-appstudio/remote-secret
# feat: validate UploadSecret keys with required key in RemoteSecret spec

### What does this PR do?
Validates the UploadSecret on two requirements:
- The type in RemoteSecret's spec and UploadSecret type have to match
- The UploadSecret has to have all the keys which are required by RemoteSecret's spec (One set of required keys is taken from the `spec.secret.keys`. The other set is inferred from the `spec.secret.type`.) 

### Screenshot/screencast of this PR

![image](https://github.com/redhat-appstudio/remote-secret/assets/73115616/21ce8305-07bf-4520-8a67-832e216766be)
![image](https://github.com/redhat-appstudio/remote-secret/assets/73115616/07c4aaf3-9fd7-41a1-a2ff-a727b0c06144)


### What issues does this PR fix or reference?

https://issues.redhat.com/browse/SVPI-535

### How to test this PR?

 1. Create RS with no required keys and uploadSecret:
 ```shell
 cat <<EOF | kubectl create -n "default" -f -
apiVersion: appstudio.redhat.com/v1beta1
kind: RemoteSecret
metadata:
  name: test-remote-secret-1
spec:
  secret:
    generateName: some-secret-
  targets: []
EOF

cat <<EOF | kubectl create -n "default" -f -
apiVersion: v1
kind: Secret
metadata:
  name: test-remote-secret-secret
  labels:
    appstudio.redhat.com/upload-secret: remotesecret
  annotations:
    appstudio.redhat.com/remotesecret-name: test-remote-secret-1
type: Opaque
stringData:
  a: b
EOF
 ```
 2. Create RS with required key `foo` and uploadSecret without `foo`. Verify that data are not uploaded, and error event is created with a message like this `message: 'validation of upload secret failed: the secret data does not contain the required keys: foo '`:
 ```shell
 cat <<EOF | kubectl create -n "default" -f -
apiVersion: appstudio.redhat.com/v1beta1
kind: RemoteSecret
metadata:
  name: test-remote-secret-2
spec:
  secret:
    keys:
    - name: foo
    generateName: some-secret-
  targets: []
EOF

cat <<EOF | kubectl create -n "default" -f -
apiVersion: v1
kind: Secret
metadata:
  name: test-remote-secret-secret
  labels:
    appstudio.redhat.com/upload-secret: remotesecret
  annotations:
    appstudio.redhat.com/remotesecret-name: test-remote-secret-2
type: Opaque
stringData:
  a: b
EOF
 ```
 3. Create RS with required key `foo` and type `kubernetes.io/ssh-auth` and uploadSecret with the same type and `foo` key. Verify that data are successfully uploaded:
 ```shell
 cat <<EOF | kubectl create -n "default" -f -
apiVersion: appstudio.redhat.com/v1beta1
kind: RemoteSecret
metadata:
  name: test-remote-secret-3
spec:
  secret:
    type: kubernetes.io/ssh-auth
    keys:
    - name: foo
    generateName: some-secret-
  targets:
  - namespace: default 
EOF

cat <<EOF | kubectl create -n "default" -f -
apiVersion: v1
kind: Secret
metadata:
  name: test-remote-secret-secret
  labels:
    appstudio.redhat.com/upload-secret: remotesecret
  annotations:
    appstudio.redhat.com/remotesecret-name: test-remote-secret-3
type: kubernetes.io/ssh-auth
stringData:
  foo: bar
  ssh-privatekey: ssh-key...
EOF
 ```
 4. Addition: Create a partial upload secret for the previous remote secret. The important part here is that the partial upload secret is type Opaque whereas the RS has ssh-auth type:
 ```shell
cat <<EOF | kubectl create -n "default" -f -
apiVersion: v1
kind: Secret
metadata:
  name: my-upload-secret
  labels:
    appstudio.redhat.com/upload-secret: remotesecret
  annotations:
    appstudio.redhat.com/remotesecret-name: test-remote-secret-3
    appstudio.redhat.com/remotesecret-partial-update: "true"
    appstudio.redhat.com/remotesecret-deleted-keys: "foo"
type: Opaque
stringData:
  ssh-privatekey: new-ssh-key...
  my-new-key: another_secret_value
EOF
 ```
 5. The remote secret should be left with:
 ```yaml
secret:                                                                                                                                                    
  keys:                                                                                                                                                    
  - my-new-key                                                                                                                                             
  - ssh-privatekey 
```
